### PR TITLE
[11][ADD] cms_account_form

### PR DESCRIPTION
**Pull request by simahawk**
**TODO**
- [x] add screenshots
- [x] decide about email update policy

### CODE REVIEW

File: cms_account_form/models/account_form.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            * provide an info message
+            * logout current session to force re-login with new email
+        """
+        email = values['email']
+        # TODO: this could be useless as we validate w/ `form_validate_email`
+        valid = validate_email(email)
+        if email and valid and not user._is_admin():
+            exists = user.sudo().search_count(
+                ['|', ('login', '=', email), ('email', '=', email)]
+            )
```
Comment (guewen) [inline cms_account_form/models/account_form.py]: Should inactive users be searched as well? I don't really know what'd happen if an inactive user exists with the same email and you don't detect it (nor how we should handle that)

File: cms_account_form/models/account_form.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            if can_change and self.o_request.website:
+                self._logout_and_notify(email)
+            return True
+        return False
+
+    def _handle_email_exists(self, email):
+        title = _('Warning')
+        msg = _(
+            'Email address `%s` already taken. '
+            'Please check inside your company. '
```
Comment (guewen) [inline cms_account_form/models/account_form.py]: Is "Please check inside your company." helpful?

**Review by guewen — COMMENTED** [review]
Looks good, 2 questions

**Comment by tschanzt** [comment]
I think that sending the Validation doesn't really do much at the moment because:
 - If I miss-type my email while changing it the validation is doesn't do anything since it is on the wrong email, which probably isn't mine.
- If someone else is changing my email without my knowledge the validation is sent to them while i don't get a hint it happened. 

Overall I think it probably would be better to send a notification to the original address with a possibility to reset it. 
Of course if you don't have access to the original any more or it was hijacked you're still gonna have a bad time. But in general it would be better imo.

Or am I missing something obvious?

File: cms_account_form/models/account_form.py (hunk 1 above)
Comment (simahawk) [reply cms_account_form/models/account_form.py]: there's an SQL constrain so it should fail badly. Probably is better if we catch that too.

File: cms_account_form/models/account_form.py (hunk 2 above)
Comment (simahawk) [reply cms_account_form/models/account_form.py]: bah, a legacy string from our project ;)

**Comment by simahawk** [comment]
@tschanzt actually right now it's worse than this: Odoo changes email + login w/out validating email nor telling anything to the user.

Hijacking: if somebody is changing your account email on the portal I guess is already too late anyway.

Sending an email to the old address: is not granted to solve the issue and is going to complicate the handling, because you'll have to:

a) post pone the update of the email
b) wait for the user to confirm the email
c) change it - if the token is still valid - when user get back

Step (b) means also that you have to implement another token machinery w/ expiration like the reset pwd does but skipping the pwd reset, of course.

So, I do know that it can be improved but from my POV it's much better to do something - like here - than nothing.

One way to make the update a bit more secure in terms of UX is to ask the user to confirm that he/she wants to really really change the email. 

What do you think?

/cc @guewen 

**EDIT:** plan B could be to decouple email from login but it means you have to change the whole registration / account behavior.

**Comment by simahawk** [comment]
@Yajo @lasley I'd like to your hear your POV about email handling. You can check the behavior in review but here is the summary of what I'm doing now:

* validate email
* check if already taken -> yes? report error - no? go to next step
* if email is changed:
  * update user's login (yes, Odoo does that w/out notifying the user!)
  * force logout
  * for reset email password to validate new email

See also my comment above.

**Review by jcoux — APPROVED** [review]
Code review only.
LGTM :+1:

**Comment by yajo** [comment]
Don't take this a review because it isn't, but seems legit.

Maybe you can leverage https://github.com/OCA/server-tools/tree/10.0/auth_signup_verify_email

**Comment by simahawk** [comment]
@Yajo you meant https://github.com/OCA/server-auth/tree/11.0/auth_signup_verify_email :wink: 

Well the point is exactly this: we are doing the same thing => reset the password. 
Nothing more, nothing special, but it's the only thing you can do ATM in Odoo w/out coding that much.

Again, my POV here is that the propose solution is in any case better than no solution as it is now in std odoo.
In the future we might implement an advanced token handling with a nice interface to handle the whole flow but is out of scope now :)

**Comment by tschanzt** [comment]
@simahawk Well it's true. It checks that we can actually send an email to the new address, which is an improvement. Maybe I was a bit to focused with what the email provides when it arrives.

I still think it would be better to notify the original email as well but if most people think it is not necessary at the moment i won't stop you.

**Review by tschanzt — APPROVED** [review]
Code is ok.

**Comment by simahawk** [comment]
@tschanzt I agree w/ you that is a "nice to have" feature: I'll add it to the roadmap session together w/ the considerations that we put on the plate ;)

**Comment by simahawk** [comment]
@guewen updated string in error message and add active_test=False to extend the search.
Also: I updated the roadmap and linked this PR + added a preview of the form

**Review by guewen — APPROVED** [review]

**Pull request merged**